lib/wlib/strings/DynamicString.cpp

```diff
@@ -0,0 +1,482 @@
+        return m_buffer[pos];
+    }
+
+    /**
+     * Element access operator gives access to character at @code pos.
+     * Character is constant
+     *
+     * @param pos the position of the character
+     * @return character at @code position
+     */
+    const char &DynamicString::operator[](size_type pos) const {
+        return m_buffer[pos];
+    }
+
+    /**
+     * Provides access to character at @code pos with bounds checking
+     *
+     * @param pos the position of the character
+     * @return character at @code position
+     */
+    char &DynamicString::at(size_type pos) {
+        return pos < len ? m_buffer[pos] : m_buffer[len];
+    }
+
+    /**
+     * Provides access to character at @code pos with bounds checking. Character is constant
+     *
+     * @param pos the position of the character
+     * @return character at @code position
+     */
+    const char &DynamicString::at(size_type pos) const {
+        return pos < len ? m_buffer[pos] : m_buffer[len];
+    }
+
+    /**
+     * Checks if string is empty or not
+     *
+     * @return if string is empty or not
+     */
+    bool DynamicString::empty() const {
+        return length() == 0;
+    }
+
+    /**
+     * Provides access to the first character in the string
+     *
+     * @return the first character
+     */
+    char &DynamicString::front() {
+        return m_buffer[0];
+    }
+
+    /**
+     * Provides access to the first character in the string. Character is constant
+     *
+     * @return the first character
+     */
+    const char &DynamicString::front() const {
+        return m_buffer[0];
+    }
+
+    /**
+     * Provides access to the last character in the string
+     *
+     * @return the last character
+     */
+    char &DynamicString::end() {
+        return empty() ? m_buffer[0] : m_buffer[len - 1];
+    }
+
+    /**
+     * Provides access to the last character in the string. Character is constant
+     *
+     * @return the last character
+     */
+    const char &DynamicString::end() const {
+        return empty() ? m_buffer[0] : m_buffer[len - 1];
+    }
+
+    /**
+     * Modifier operator adds character to the current string.
+     *
+     * @param c character to add
+     * @return the current string
+     */
+    DynamicString &DynamicString::operator+=(char c) {
+        char array[1] = {c};
+        return append(array);
+    }
+
+    /**
+     * Modifier operator adds char string to the current string.
+     *
+     * @param val char string to add
+     * @return the current string
+     */
+    DynamicString &DynamicString::operator+=(const char *val) {
+        return append(val);
+    }
+
+    /**
+     * Modifier operator adds @code DynamicString object to the current string.
+     *
+     * @param other @code DynamicString string to add
+     * @return the current string
+     */
+    DynamicString &DynamicString::operator+=(DynamicString &other) {
+        return append(other.c_str());

```

Comment:
I would make the function `append(const char *cstr, size_type len)` and have overloads for `append(const DynamicString &str)` which calls `append(str.c_str(), str.length())` and `append(const char *str)` which calls `append(str, strlen(str))`

This avoids computing the length of the `c_str()` value of a dynamic string when the length is already known.